src/dualized_block.jl

```diff
@@ -0,0 +1,78 @@
+"""
+    DualizedNLPblockModel{T, S} <: AbstractNLPModel{T, S}   
+
+A data type to store dualized subproblems. 
+"""
+mutable struct DualizedNLPblockModel{T, S} <: AbstractNLPModel{T, S}
+  meta::NLPModelMeta{T, S}
+  counters::Counters
+  problem_block::AbstractNLPModel
+  λ::AbstractVector # dual variables
+  A::AbstractArray # linking matrices
+end
+
+"""
+    DualizedNLPblockModel(
+      nlp::AbstractNLPModel, 
+      λ::AbstractVector, 
+      A::AbstractArray
+      )
+
+Modifies a subproblem by dualizing the linking constraints.
+
+# Arguments
+- `nlp::AbstractNLPModel`: the subproblem 
+- `λ::AbstractVector`: vector of dual variables
+- `A::AbstractArray`: block linking matrix corresponding to the subproblem `nlp`
+"""
+function DualizedNLPblockModel(nlp::AbstractNLPModel, λ::AbstractVector, A::AbstractArray)
+  meta = nlp.meta
+  return DualizedNLPblockModel(meta, Counters(), nlp, λ, A)
+end
+
+function NLPModels.obj(nlp::DualizedNLPblockModel, x::AbstractVector)
+  n = nlp.problem_block.meta.nvar
+  d(x) = dot(nlp.λ, nlp.A, x)
+  return obj(nlp.problem_block, x) + d(x)
+end
+
+function NLPModels.grad!(nlp::DualizedNLPblockModel, x::AbstractVector, g::AbstractVector)
+  grad!(nlp.problem_block, x, g)
+  g .+= nlp.A'*nlp.λ
```

Comment:
I would replace this line by using `mul!` directly:
```julia
mul!(g, nlp.A', nlp.λ, 1, 1)
```

Reply:
Won't this overwrite the values stored in `g`? Whereas what I want to do here is to add `nlp.A'*nlp.λ` to the existing elements of `g`.

Reply:
The `mul!` syntax is a bit misleading in Julia. `mul!` with three arguments does overwrite, but `mul!` with 5 arguments modifies the values inplace. See the documentation.